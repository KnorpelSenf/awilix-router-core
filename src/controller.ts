import {
  MiddlewareParameter,
  createState,
  addHttpVerbs,
  getOrInitMethodConfig,
  addRoute,
  addBeforeMiddleware,
  addAfterMiddleware
} from './state-util'
import { HttpVerbs } from './http-verbs'
import { HttpVerb, Constructor } from 'src'
import { STATE, IS_CONTROLLER_BUILDER } from './symbols'

/**
 * Method builder options.
 */
export interface IMethodBuilderOpts {
  before?: MiddlewareParameter
  after?: MiddlewareParameter
}

/**
 * Verb builder function.
 */
export type VerbBuilderFunction = (
  path: string,
  method: string,
  opts?: IMethodBuilderOpts
) => IAwilixControllerBuilder

/**
 * Fluid router builder.
 */
export interface IAwilixControllerBuilder {
  target: Constructor | Function
  get: VerbBuilderFunction
  post: VerbBuilderFunction
  put: VerbBuilderFunction
  patch: VerbBuilderFunction
  delete: VerbBuilderFunction
  head: VerbBuilderFunction
  options: VerbBuilderFunction
  connect: VerbBuilderFunction
  all: VerbBuilderFunction
  verbs(
    verbs: Array<HttpVerb>,
    path: string,
    method: string,
    opts?: IMethodBuilderOpts
  ): this
  prefix(path: string): this
  before(value: MiddlewareParameter): this
  after(value: MiddlewareParameter): this
}

/**
 * Configures routing config for a class or function to be invoked by a router.
 *
 * @example
 *   const api = ({ todoService }) => ({
 *     find: (ctx) => { ... }
 *   })
 *
 *   export default createController(api)
 *     .prefix('/todos')
 *     .before(bodyParser())
 *     .get('/', 'find')
 *     .get('/:id', 'get')
 *     .post('/:id', 'create', {
 *       before: [authenticate()],
 *     })
 *     .patch('/:id', 'update', {
 *       before: [authenticate()]
 *     })
 */
export function createController(
  ClassOrFunction: Constructor | Function
): IAwilixControllerBuilder {
  const state = createState()
  const builder: IAwilixControllerBuilder = {
    [STATE]: state,
    [IS_CONTROLLER_BUILDER]: true,
    target: ClassOrFunction,
    get: createVerbFunction(HttpVerbs.GET),
    post: createVerbFunction(HttpVerbs.POST),
    put: createVerbFunction(HttpVerbs.PUT),
    patch: createVerbFunction(HttpVerbs.PATCH),
    delete: createVerbFunction(HttpVerbs.DELETE),
    head: createVerbFunction(HttpVerbs.HEAD),
    options: createVerbFunction(HttpVerbs.OPTIONS),
    connect: createVerbFunction(HttpVerbs.CONNECT),
    all: createVerbFunction(HttpVerbs.ALL),
    prefix(path) {
      addRoute(state.root, path)
      return builder
    },
    before(middleware) {
      addBeforeMiddleware(state.root, middleware)
      return builder
    },
    after(middleware) {
      addAfterMiddleware(state.root, middleware)
      return builder
    },
    verbs(verbs, path, method, opts) {
      const methodConfig = getOrInitMethodConfig(state, method)
      addRoute(methodConfig, path)
      addHttpVerbs(methodConfig, verbs)
      if (opts) {
        if (opts.before) {
          addBeforeMiddleware(methodConfig, opts.before)
        }

        if (opts.after) {
          addAfterMiddleware(methodConfig, opts.after)
        }
      }
      return builder
    }
  }

  return builder

  /**
   * Creates a preconfigured verb function.
   *
   * @param verb
   */
  function createVerbFunction(verb: HttpVerb): VerbBuilderFunction {
    return function configureRoute(path, method, opts) {
      return builder.verbs([verb], path, method, opts)
    }
  }
}
