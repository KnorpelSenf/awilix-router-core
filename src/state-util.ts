import { HttpMethod } from './http-methods'

/**
 * Symbol used for getting and setting the config state.
 */
const STATE = Symbol('Router Config')

/**
 * Router config state.
 */
export interface IRouterConfigState {
  /**
   * Root config (class-level).
   */
  root: IRouteConfig
  /**
   * Method configs (method-level).
   */
  methods: Map<string, IRouteConfig>
}

/**
 * A specific route config.
 */
export interface IRouteConfig {
  /**
   * Paths to register.
   */
  paths: Array<string>
  /**
   * Middleware to run before the method.
   */
  beforeMiddleware: Array<any>
  /**
   * Middleware to run after the method.
   */
  afterMiddleware: Array<any>
  /**
   * HTTP methods to register.
   */
  methods: Array<HttpMethod>
}

/**
 * Gets or initializes a method config.
 *
 * @param state
 * @param name
 */
export function getOrInitMethodConfig(state: IRouterConfigState, name: string) {
  const config = state.methods.get(name)
  if (!config) {
    const newConfig = createRouteConfig()
    state.methods.set(name, newConfig)
    return newConfig
  }

  return config
}

/**
 * Gets the config state from the target.
 *
 * @param target
 */
export function getState(target: any): IRouterConfigState | null {
  return (target.prototype ? target.prototype[STATE] : target[STATE]) || null
}

/**
 * Sets the config state on the target.
 * @param target
 * @param state
 */
export function setState(target: any, state: IRouterConfigState) {
  if (target.prototype) {
    target.prototype[STATE] = state
  } else {
    target[STATE] = state
  }
  return state
}

/**
 * Gets or initializes the configuration for a decorator.
 * If it's a method decorator, we initialize and return one of those,
 * else we return the root config.
 *
 * @param target
 * @param name
 */
export function getOrInitConfigForDecorator(target: any, name?: string) {
  const state = getState(target) || setState(target, createState())
  const config = name ? getOrInitMethodConfig(state, name) : state.root
  return config
}

/**
 * Creates a new state object.
 */
export function createState(): IRouterConfigState {
  const state: IRouterConfigState = {
    root: createRouteConfig(),
    methods: new Map<string, IRouteConfig>()
  }
  return state
}

/**
 * Creates a new route config object.
 */
export function createRouteConfig(): IRouteConfig {
  return {
    paths: [],
    beforeMiddleware: [],
    afterMiddleware: [],
    methods: []
  }
}

/**
 * Rolls up state so paths are joined, middleware rolled into
 * the correct order, etc.
 *
 * @param state
 */
export function rollUpState(
  state: IRouterConfigState
): Map<string, IRouteConfig> {
  const result = new Map<string, IRouteConfig>()
  state.methods.forEach((method, key) => {
    result.set(key, {
      paths: concatPaths(state.root.paths, method.paths),
      beforeMiddleware: [
        ...state.root.beforeMiddleware,
        ...method.beforeMiddleware
      ],
      afterMiddleware: [
        ...method.afterMiddleware,
        ...state.root.afterMiddleware
      ],
      methods: method.methods
    })
  })
  return result
}

/**
 * Concatenates root and method paths so we have one for each combination.
 */
function concatPaths(rootPaths: Array<string>, methodPaths: Array<string>) {
  const result: Array<string> = []
  rootPaths.forEach(rootPath => {
    methodPaths.forEach(methodPath => {
      result.push(rootPath + methodPath)
    })
  })

  return result
}
