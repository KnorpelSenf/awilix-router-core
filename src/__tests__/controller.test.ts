import { createController } from '../controller';
import { getState, rollUpState } from '../state-util';

describe('createController', () => {
  it('creates the correct state', () => {
    const target = () => {
      const noop = () => undefined;
      return { all: noop, get: noop, post: noop, put: noop };
    };
    const router = createController(target)
      .prefix('/root1')
      .prefix('/root2')
      .before('beforeRoot1')
      .before('beforeRoot2')
      .after('afterRoot1')
      .after('afterRoot2')
      .all('/all', 'all', { before: ['beforeAll'] })
      .get('/get', 'get', { before: ['beforeGet'], after: 'afterGet' })
      .post('/post', 'post')
      .put('/', 'put', {});
    const state = rollUpState(getState(router)!);
    expect(state).toMatchSnapshot();
  });
});
