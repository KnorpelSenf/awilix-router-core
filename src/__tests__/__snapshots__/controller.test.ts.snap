// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createController creates the correct state 1`] = `
Map {
  "all" => Object {
    "afterMiddleware": Array [
      "afterRoot1",
      "afterRoot2",
    ],
    "beforeMiddleware": Array [
      "beforeRoot1",
      "beforeRoot2",
      "beforeAll",
    ],
    "paths": Array [
      "/root1/all",
      "/root2/all",
    ],
    "verbs": Array [
      "*",
    ],
  },
  "get" => Object {
    "afterMiddleware": Array [
      "afterGet",
      "afterRoot1",
      "afterRoot2",
    ],
    "beforeMiddleware": Array [
      "beforeRoot1",
      "beforeRoot2",
      "beforeGet",
    ],
    "paths": Array [
      "/root1/get",
      "/root2/get",
    ],
    "verbs": Array [
      "GET",
    ],
  },
  "post" => Object {
    "afterMiddleware": Array [
      "afterRoot1",
      "afterRoot2",
    ],
    "beforeMiddleware": Array [
      "beforeRoot1",
      "beforeRoot2",
    ],
    "paths": Array [
      "/root1/post",
      "/root2/post",
    ],
    "verbs": Array [
      "POST",
    ],
  },
  "put" => Object {
    "afterMiddleware": Array [
      "afterRoot1",
      "afterRoot2",
    ],
    "beforeMiddleware": Array [
      "beforeRoot1",
      "beforeRoot2",
    ],
    "paths": Array [
      "/root1/",
      "/root2/",
    ],
    "verbs": Array [
      "PUT",
    ],
  },
}
`;
